netcore/System.Private.CoreLib/src/System/Environment.Unix.cs

```diff
@@ -13,36 +13,91 @@ namespace System
 {
 	partial class Environment
 	{
+		private static Dictionary<string, string> s_environment;
+
 		static string GetEnvironmentVariableCore (string variable)
 		{
-			using (var h = RuntimeMarshal.MarshalString (variable)) {
-				return internalGetEnvironmentVariable_native (h.Value);
+			Debug.Assert(variable != null);
+
+			if (s_environment == null) {
+				using (var h = RuntimeMarshal.MarshalString (variable)) {
+					return internalGetEnvironmentVariable_native (h.Value);
+				}
+			}
+
+			lock (s_environment) {
+				variable = TrimStringOnFirstZero (variable);
+				s_environment.TryGetValue (variable, out string value);
+				return value;
 			}
 		}
 
-		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		extern static string internalGetEnvironmentVariable_native (IntPtr variable);
+		static unsafe void SetEnvironmentVariableCore (string variable, string value)
+		{
+			Debug.Assert(variable != null);
 
-		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		private extern static string [] GetEnvironmentVariableNames ();
+			EnsureEnvironmentCached ();
+			lock (s_environment) {
+				variable = TrimStringOnFirstZero (variable);
+				value = value == null ? null : TrimStringOnFirstZero (value);
+				if (string.IsNullOrEmpty (value)) {
+					s_environment.Remove (variable);
+				} else {
+					s_environment[variable] = value;
+				}
+			}
+		}
 
 		public static IDictionary GetEnvironmentVariables ()
 		{
-			Hashtable vars = new Hashtable ();
-			foreach (string name in GetEnvironmentVariableNames ()) {
-				vars [name] = GetEnvironmentVariableCore (name);
+			var results = new Hashtable();
+
+			EnsureEnvironmentCached();
+			lock (s_environment) {
+				foreach (var keyValuePair in s_environment) {
+					results.Add(keyValuePair.Key, keyValuePair.Value);
+				}
 			}
-			return vars;
+
+			return results;
 		}
 
-		static unsafe void SetEnvironmentVariableCore (string variable, string value)
+		private static string TrimStringOnFirstZero (string value)
 		{
-			fixed (char *fixed_variable = variable)
-			fixed (char *fixed_value = value)
-				InternalSetEnvironmentVariable (fixed_variable, variable.Length, fixed_value, value?.Length ?? 0);
+			int index = value.IndexOf('\0');
```

Comment:
Follow the coding style with space before `(`

Reply:
Fixed.